src/main/protobuf/game.proto

```diff
@@ -2,19 +2,237 @@ syntax = "proto3";
+ *     3: {gain: [], lose: [ORE, ORE]},
+ *     4: {gain: [] lose: []},
+ *   message: null
+ * }
+ * GameEvent {
+ *   position: 3, 
+ *   action: BUILD_CITY,
+ *   specification: {vertex: B4R},
+ *   result: {resources_transacted: {
+ *     1: {gain: [], lose: []},
+ *     2: {gain: [], lose: []},
+ *     3: { gain: [], lose: [ORE, ORE, ORE, WHEAT, WHEAT]},
+ *     4: {gain: [] lose: []},
+ *   message: null
+ *  }
+ * Important to note, is that multiple actions are able to recycle specification.
+ * For example, BUILD_CITY and BUILD_ROAD have the same specification.
+ * Similarly proposing a trade, making a trade, playing a monopoly or a year of plenty
+ * all take advantage of one or both of the `ask` and `give` fields in the ActionSpecification.
+ **/
+message GameEvent {
+  int32 position = 1; // Player positions on the server need to be 1 not 0 based, so 0 can indicate no position, or we can pass -1
+  GameAction action = 2;
+  ActionSpecification specification = 3;
+  ActionResult result = 4;
+  string message = 5;
+}
+
+/**
+ * Defines the action a player wants to make.
+ * Is used both in the TakeAction RPC and in the GameEvent response.
+ * Of note, like the GameAction enum, the ActionSpecification object used in the TaekAction RPC
+ * is directly piped through into the GameEvent response.
+ * The GameEvent response might optionally also inlcude a result.
+ * For more inforamtion see GameEvent.
+ * Depending on the action being taken, different combinations of the below fields will be specified.
+ * See mapping below:
+ * GAME_ACTION_NONE --> No Action Spec
+ * INITIAL_PLACEMENT --> vertex, edge
+ * ROLL_DICE --> No Action Spec
+ * MOVE_ROBBER_AND_STEAL --> hex, other_player
+ * BUILD_ROAD --> edge
+ * BUILD_SETTLEMENT --> vertex
+ * BUILD_CITY --> vertex
+ * BUILD_DEVELOPMENT_CARD --> No Action Spec
+ * ACTIVATE_KNIGHT --> hex, other_player
+ * ACTIVATE_ROAD_BUILDING --> edges
+ * ACTIVATE_YEAR_OF_PLENTY --> ask
+ * ACTIVATE_MONOPOLY --> ask
+ * ACTIVATE_VICTORY_POINT --> No Action Spec
+ * PROPOSE_TRADE --> ask, give
+ * PORT_TRADE --> ask, give
+ * ACCEPT_TRADE --> ask, give
+ * REJECT_TRADE --> ask, give
+ * DISCARD --> give
+ * END_TURN --> No Action Spec
+ * ACKNOWLEDGE --> No Action Spec
+ **/
+message ActionSpecification {
+  string hex = 2;
+  string vertex = 3;
+  Edge edge = 4;
+  repeated Edge edges = 5; // Road Building
+  int32 other_player_position = 7; // Trade, Steal
+  repeated Resource ask = 8; // Trade, Monopoly, Year of Plenty
+  repeated Resource give = 9; // Trade, Discard
+}
+
+message Edge {
+  string v1 = 1;
+  string v2 = 2;
+}
+
+/**
+ * Specifies the result of a TakeAction RPC call.
+ * Used as a subfield of the GameEvent response.
+ * Dependingon the GameAction, there is a different payload, as specified in the mapping below
+ * GAME_ACTION_NONE, INITIAL_PLACEMENT, ACTIVATE_VICTORY_POINT, PROPOSE_TRADE, REJECT_TRADE, END_TURN, ACKNOWLEDGE   --> No Action Result
```

Comment:
I dont think ACTIVATE_VICTORY_POINT is an action a player should be able to perform. They have the points they cant activate them. Once the game is over their points are automatically activated

Reply:
I actually disagree. First, because I think actions should be client driven (I changed my mind on that--originally I thought the server should "automatically" roll the dice for you if you don't have a knight to play or chose not to play one--now I think you should have to tell the server you want to roll the dice).

Second, I don't know the specific Catan rules, but I think, if you are dumb, you can forget to reveal your VPs and lose, right?

Further, you might be able to reveal them early? (or is that technically against the rules?)

Reply:
Im not sure the exact rules but this goes to my philosophy of the game. I believe that when you have a VP card, you have the point and the card is never actually played. The game ends as soon as someone has at least 10 points. Other people think you need to claim victory. I do not think so. The way we are programming the game supports this because when a player pick up a VP the point is automatically added to the players counter and the game ends as soon as someone has at least 10 points. 
I am pretty sure you cannot show your development cards until you play them so if you can't play the VP cards until you win I actually do think it is against the rules to show your card. 
That is how I think of the game. I do not want to give the client the ability to reveal their VP cards I think that is a ridiculous action and I have never seen an online version of catan allow a player to do that

Reply:
Okay, sounds good to me 👍